### ISSUE TITLE
[cssom] The way CSSStyleDeclaration exposes properties is not ideal.

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#5649]
Right now we only have a single interface that exposes the CSS property names as attributes ([`CSSStyleDeclaration`](https://drafts.csswg.org/cssom/#the-cssstyledeclaration-interface)), but no browser does it right, and the spec behavior is also not great.

In particular, the spec behavior means that you expose stuff like `fontFaceRule.style.color`, which doesn't quite make sense.

WebKit and Blink (though WebKit may have changed recently) historically use named getters for this (which is also not great, see https://github.com/w3c/csswg-drafts/issues/1089).

Note also that:

 * Per recent discussion with @gsnedders WebKit trunk may have changed behavior?
 * Per spec there's no exposure of descriptors for the rules that have them (but Blink / WebKit expose them as part of the named getters).

Gecko does something which IMO is better. Gecko exposes them as attributes as the spec describes, but on an interface that derives from `CSSStyleDeclaration` called [`CSS2Properties`](https://searchfox.org/mozilla-central/rev/25d5a4443a7e13cfa58eff38f1faa5e69f0b170f/dom/webidl/CSS2Properties.webidl.in#2) which we use for all the "style" declarations (so, inline style declarations, `CSSStyleRule.style`, `getComputedStyle`, `CSSPageRule.style`, and individual keyframe declarations). ([link](https://searchfox.org/mozilla-central/rev/25d5a4443a7e13cfa58eff38f1faa5e69f0b170f/layout/style/nsDOMCSSDeclaration.cpp#31), [link](https://searchfox.org/mozilla-central/search?q=public+nsDOMCSSDec&path=&case=false&regexp=false)). I think something like what Gecko does, but with similar interfaces for the declaration blocks that should have descriptors, like `@font-face` and `@counter-style`, would generally be much saner.

Since there's movement around here, and the interop story here is quite sad, I think there's quite a fair amount of room for improvement, and if we can align in a model that is both consistent and useful for developers it'd be great.

Does the thing I described above seem reasonable? Other ideas?

cc @gsnedders @foolip @andruud @xiaochengh @heycam

**Comment by gsnedders** [comment]
> Per recent discussion with @gsnedders WebKit trunk may have changed behavior?

WebKit trunk basically follows the CSSOM spec here, with each property (and descriptor, because we didn't stop exposing them) a separate IDL attribute rather than using named getters. The old named getter had some bugs that led to some "interesting" behaviour hence that change.

**Comment by emilio** [comment]
I could probably live with that too, if people think it's fine to expose something like `'src' in document.body.style`.

**Comment by emilio** [comment]
(Though not a terrible fan of it it might be the path of least resistance...)

I assume that for stuff that doesn't apply you just always return the empty string?

**Comment by gsnedders** [comment]
I don't think we have any opposition to changing behaviour for descriptors; it was just the simple change to move away from the named getter (and also effectively the minimal change).

**Comment by andruud** [comment]
@emilio Your proposal sounds good to me. Having to expose `src` and other descriptors on `.style` (etc) feels like a design failure, so it's a bit annoying. (Though maybe not a huge problem for authors in practice).

**Comment by tabatkins** [comment]
Yes, I've been wanting to improve this for a long time; the current situation is ridiculous.

Having child interfaces for properties and for each at-rule that contains only the properties/descriptors relevant to the context would be great.

**Comment by LeaVerou** [comment]
Just a note to consider introspection for whatever is decided, which is immensely useful for tooling. The situation right now is [far from ideal](https://lea.verou.me/2020/07/introspecting-css-via-the-css-om-getting-supported-properties-shorthands-longhands/).

**Comment by argyleink** [comment]
issues related to this were the hardest part of getting VisBug to work across browser

**Comment by emilio** [comment]
So, here's a proposal. All of the interfaces below would extend `CSSStyleDeclaration` and thus have a `getPropertyValue` / `setProperty` / `cssText` / `length` / `item` / etc...:

 * `CSSStyleProperties` (`CSS2Properties`) for style declarations: inline style / `getComputedStyle` / `CSSStyleRule.style`. This is basically the current `CSSStyleDeclaration`. I like `CSSStyleProperties` a bit more, because it's more consistent, but there's precedent for `CSS2Properties`, and that's what Gecko calls it, so it's a bit more compatible.
 * `CSSPageProperties` for `@page` rules, with only the page descriptor.
 * `CSSCounterStyleProperties` for `@counter-style`
 * `CSSFontFaceProperties` for `@font-face`

etc. Two thoughts:

 * I'd prefer to use `CSS{Style,Page,...}Declarations`, but I think that might be confusing with the existing `CSSStyleDeclaration`.
 * An alternative that doesn't use `Properties` would be `Block` (so, `CSSStyleBlock` / `CSSPageBlock` / ...), which could be nice because it avoids using `Properties` for something that's "descriptors"...

But this eventually comes down to naming, so... thoughts?

**Comment by gsnedders** [comment]
> `CSSStyleProperties` (`CSS2Properties`) for style declarations: inline style / `getComputedStyle` / `CSSStyleRule.style`. This is basically the current `CSSStyleDeclaration`. I like `CSSStyleProperties` a bit more, because it's more consistent, but there's precedent for `CSS2Properties`, and that's what Gecko calls it, so it's a bit more compatible.

That precedent comes from [DOM Level 2 Style](https://www.w3.org/TR/DOM-Level-2-Style/css.html#CSS-extended). From an aesthetic point-of-view I'd like to get rid of the lie that it's only CSS2, so if you wanna try renaming it in Gecko and see if it sticks I'd quite like that?

> An alternative that doesn't use `Properties` would be `Block` (so, `CSSStyleBlock` / `CSSPageBlock` / ...), which could be nice because it avoids using `Properties` for something that's "descriptors"...

Is there any reason why we can't go for `CSSStyleProperties`/`CSS2Properties` then `CSSPageDescriptors`, `CSSCounterStyleDescriptors`, etc?

**Comment by emilio** [comment]
I guess not really, that'd be fine for me.

**Comment by svgeesus** [comment]
>> Is there any reason why we can't go for CSSStyleProperties/CSS2Properties then CSSPageDescriptors, CSSCounterStyleDescriptors, etc?

> I guess not really, that'd be fine for me.

I'm not following what specific change I should make to [CSS Fonts 4 @font-face interface](https://drafts.csswg.org/css-fonts-4/#om-fontface) for this.

**Comment by tabatkins** [comment]
You'd make a `CSSFontFaceDescriptors` interface, containing the `@font-face` descriptors, and then change the `CSSFontFaceRule.style` attribute to be of that type (rather than `CSSStyleDeclaration`).

**Comment by gsnedders** [comment]
Looking at this again, can anyone make any of the exposed attributes for the at-rules which expose a `CSSStyleDeclaration` do… anything? If it truly is the case that "setting" a descriptor like this does nothing in WebKit/Chromium then we should probably just drop them entirely, which means we only need `CSSStyleDeclaration` and `CSSStyleProperties`.

**Comment by cdoublev** [comment]
I went through issues related to this one, and read the following from a meeting transcript:

> <dino> jdaggett: are people actually using this functionality, that's been there since CSS 2. Last time I looked, if you went in and changed the font family, font matching didn't respond correctly.
> <dino> myles_: works in Safari.
> <dino> jdaggett: not in Chrome

But I guess attributes for at-rules do nothing if you say so and if I understood you correctly.

Regarding the naming that seems to be preferred for now, is `CSSStyleProperties.cssText = & { color: green }` supposed to work? If so, I would say that something like `CSSStyleBlock[Value|Content]` makes more sense to me.

**Comment by emilio** [comment]
@gsnedders Setting a descriptor "works" for some rules, like `@page` at least.

**Comment by emilio** [comment]
That's basically https://drafts.csswg.org/css-fonts-4/#om-fontpalettevalues right?

**Comment by saschanaz** [comment]
Can this be closed now as https://github.com/w3c/csswg-drafts/pull/9686 is merged?
